start of 'data manipulation' examples

bcoca opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Docs Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
docs
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
2.x
```

bcoca committed:
start of 'data manipulation' examples

abadger commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,71 @@
+
+A quick definition of each type of plugin before we go forward
+  - lookups: Mainly used to query 'external data', in Ansible these were the primary part of loops using the `with_<lookup>` construct, but they can be used independantly to return data for processing. They normally returna list due to their primary function in loops as mentioned previouslly. Used with the ``lookup`` or ``query`` Jinja2 operators.
+  - filters: used to change/transform data, used mainly with the ``|`` Jinja2 operator.
```
s/mainly //

bcoca replied:
map/select as the 'other usage', so they are not exclusively used with | or is

abadger commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,71 @@
+A quick definition of each type of plugin before we go forward
+  - lookups: Mainly used to query 'external data', in Ansible these were the primary part of loops using the `with_<lookup>` construct, but they can be used independantly to return data for processing. They normally returna list due to their primary function in loops as mentioned previouslly. Used with the ``lookup`` or ``query`` Jinja2 operators.
+  - filters: used to change/transform data, used mainly with the ``|`` Jinja2 operator.
+  - tests: used to validate data, used mainly with the ``is`` Jinja2 operator.
```
I think s/mainly //  (Is there an alternate way to use tests?)

abadger replied:
Probably should have a seealso for each of these three plugin types to the list of plugins of that type

bcoca replied:
^ responded above, map is alternate for filters select/reject for tests

abadger commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,71 @@
+Most programmers are used to using a for loop (or while and others) and list comprehensions to do transformations on lists or lists of objects, Jinja2 has a few filters that provide this functionality: map, select, reject, selectattr, rejectattr
+
+ - map: this is a basic for loop that just allows you to transform every item in a list, using the 'attribute' keyword you can do the transformation based on a attribute of the list element.
+
```
May need to list the Jinja2 minimum version here? (Or do we not have to worry about that because we're no longer supporting installing on the RHEL6 python-2.6 stack?)

bcoca replied:
added note

abadger commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,71 @@
+========
+
+.. _keys_from_dict_matching_list:
+Extract keys form a dictionary matching elements from a list
```
s/form/from/

abadger commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,71 @@
+        for each config in chains_config[chain]['configs']
+            print config['type']
+        end for
+    end for
```
I'd start with the pseudo-code since it expresses the idea that people will want to translate from.

Using python instead of arbitrary pseudo code is probably better.  For instance, we can get the code syntax highlighted and people can consult python documentation (or run it in a python shell) if they're confused about a nuance of the example.

bcoca committed:
some updates from feedback

 - draft 'get mount of path'
 - added list selection and mount example
 - added 'dict fun'
 - updated iwth omited lists

Co-Authored-By: Klaus Frank <agowa338@users.noreply.github.com>
Co-Authored-By: Felix Fontein <felix@fontein.de>

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

abadger commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,83 @@
+        for each config in chains_config[chain]['configs']
+            print config['type']
+        end for
+    end for
```
```suggestion
This recipe performs the equivalent of this pseudo-code:

.. code-block:: python
    for chain in chains:
        for config in chains_config[chain]['configs']:
            print(config['type'])
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+Complex data manipulation
+#########################
+
+In many cases you need to do some complex operation with your variables, while Ansible is not recommended as a data processing/manipulation tool, you can use the existing Jinja2 templating in conjunction with the many added ansible filters, lookups and tests to do some very complex transformations.
```
```suggestion
In many cases you need to do some complex operation with your variables. Although Ansible is not recommended as a data processing/manipulation tool, you can use the existing Jinja2 templating in conjunction with the many added Ansible filters, lookups and tests to do some very complex transformations.
```

gundalow replied:
```suggestion
In many cases, you need to do some complex operation with your variables, while Ansible is not recommended as a data processing/manipulation tool, you can use the existing Jinja2 templating in conjunction with the many added Ansible filters, lookups and tests to do some very complex transformations.
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+
+In many cases you need to do some complex operation with your variables, while Ansible is not recommended as a data processing/manipulation tool, you can use the existing Jinja2 templating in conjunction with the many added ansible filters, lookups and tests to do some very complex transformations.
+
+A quick definition of each type of plugin before we go forward
```
```suggestion
Let's start with a quick definition of each type of plugin:
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+In many cases you need to do some complex operation with your variables, while Ansible is not recommended as a data processing/manipulation tool, you can use the existing Jinja2 templating in conjunction with the many added ansible filters, lookups and tests to do some very complex transformations.
+
+A quick definition of each type of plugin before we go forward
+  - lookups: Mainly used to query 'external data', in Ansible these were the primary part of loops using the `with_<lookup>` construct, but they can be used independantly to return data for processing. They normally returna list due to their primary function in loops as mentioned previouslly. Used with the ``lookup`` or ``query`` Jinja2 operators.
```
```suggestion
  - lookups: mainly used to query 'external data', and in loops using the ``with_<lookup>`` construct. Lookup plugins can also be used independently to return data for processing. They normally return a list due to their primary function in loops. Used with the ``lookup`` or ``query`` Jinja2 operators.
```

gundalow replied:
```suggestion
  - lookups: Mainly used to query 'external data', in Ansible these were the primary part of loops using the ``with_<lookup>`` construct, but they can be used independently to return data for processing. They normally return a list due to their primary function in loops as mentioned previously. Used with the ``lookup`` or ``query`` Jinja2 operators.
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+
+.. _for_loops_or_list_comprehensions:
+
+For loops and/or list comprehensions
```
```suggestion
``For`` loops and/or list comprehensions
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+For loops and/or list comprehensions
+====================================
+
+Most programmers are used to using a for loop (or while and others) and list comprehensions to do transformations on lists or lists of objects, Jinja2 has a few filters that provide this functionality: map, select, reject, selectattr, rejectattr
```
```suggestion
Most programmers are used to using a ``for`` loop (or ``while``, etc.) and list comprehensions to do transformations on lists or lists of objects, Jinja2 has a few filters that provide this functionality: ``map``, ``select``, ``reject``, ``selectattr``, and ``rejectattr``.
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+
+Most programmers are used to using a for loop (or while and others) and list comprehensions to do transformations on lists or lists of objects, Jinja2 has a few filters that provide this functionality: map, select, reject, selectattr, rejectattr
+
+ - map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on a attribute of the list element.
```
```suggestion
 - map: this is a basic ``for`` loop that allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on an attribute of the list element.
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+
+ - map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on a attribute of the list element.
+
+ - select/reject: this is a for loop with a conditional, that allows you to create a subset of a list that matches (or not) a condition
```
```suggestion
 - select/reject: this is a ``for`` loop with a conditional, that allows you to create a subset of a list that matches (or not) a condition.
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+ - map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on a attribute of the list element.
+
+ - select/reject: this is a for loop with a conditional, that allows you to create a subset of a list that matches (or not) a condition
+ - selectattr/rejectattr: very similar to the above but it uses a specific attribute of the list element for the conditional
```
```suggestion
 - selectattr/rejectattr: very similar to select/reject, but using a specific attribute of the list element for the conditional.
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+Complex data manipulation
```
```suggestion
.. _complex_data_manipulation:

*************************
Complex data manipulation
```

acozine commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,103 @@
+Complex data manipulation
+#########################
```
```suggestion
*************************
```

acozine reviewed (commented):
This is great content, thanks @bcoca! I've done a bit of wordsmithing on it and suggested some "mechanical" fixes in line with https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html.

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+In many cases, you need to do some complex operation with your variables, while Ansible is not recommended as a data processing/manipulation tool, you can use the existing Jinja2 templating in conjunction with the many added Ansible filters, lookups and tests to do some very complex transformations.
+
+A quick definition of each type of plugin before we go forward
+  - lookups: Mainly used to query 'external data', in Ansible these were the primary part of loops using the ``with_<lookup>`` construct, but they can be used independently to return data for processing. They normally returna a list due to their primary function in loops as mentioned previously. Used with the ``lookup`` or ``query`` Jinja2 operators.
```
```suggestion
  - lookups: Mainly used to query 'external data', in Ansible these were the primary part of loops using the ``with_<lookup>`` construct, but they can be used independently to return data for processing. They normally return a list due to their primary function in loops as mentioned previously. Used with the ``lookup`` or ``query`` Jinja2 operators.
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+For loops and/or list comprehensions
+====================================
+
+Most programmers are used to using a ``for`` loop (or ``while``, etc) and list comprehensions to do transformations on lists or lists of objects, Jinja2 has a few filters that provide this functionality: ``map``, ``select``, ``reject``, ``selectattr``, ``rejectattr``.
```
```suggestion
Most programmers are used to use loops (``for``, ``while``, etc) and list comprehensions to do transformations on lists including lists of objects, Jinja2 has a few filters that provide this functionality: ``map``, ``select``, ``reject``, ``selectattr``, ``rejectattr``.
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+
+Most programmers are used to using a ``for`` loop (or ``while``, etc) and list comprehensions to do transformations on lists or lists of objects, Jinja2 has a few filters that provide this functionality: ``map``, ``select``, ``reject``, ``selectattr``, ``rejectattr``.
+
+  - map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on a attribute of the list element.
```
```suggestion
  - map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on attributes of the list elements.
```
or
```suggestion
  - map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on an attribute of the list elements.
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+
+  - map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on a attribute of the list element.
+
+  - select/reject: this is a for loop with a conditional, that allows you to create a subset of a list that matches (or not) a condition
```
```suggestion
  - select/reject: this is a for loop with a condition, that allows you to create a subset of a list that matches (or not) based on the result of the condition.
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+  - map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on a attribute of the list element.
+
+  - select/reject: this is a for loop with a conditional, that allows you to create a subset of a list that matches (or not) a condition
+  - selectattr/rejectattr: very similar to the above but it uses a specific attribute of the list element for the conditional
```
```suggestion
  - selectattr/rejectattr: very similar to the above but it uses a specific attribute of the list elements for the conditional statement.
```
or
```suggestion
  - selectattr/rejectattr: very similar to the above but it uses a specific attribute of the list elements for the condition.
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+Complex Type transformations
+=============================
+
+Jinja provides filters for simple data type tranformations (int, bool, etc), but when you want to tranform data structures things are not as easy.
```
```suggestion
Jinja provides filters for simple data type transformations (int, bool, etc), but when you want to transform data structures things are not as easy.
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+=============================
+
+Jinja provides filters for simple data type tranformations (int, bool, etc), but when you want to tranform data structures things are not as easy.
+You can use the loops and list comprehensions as shown above to help, also other filters and lookups can be chained and leverage to achieve these comlex transformations.
```
```suggestion
You can use loops and list comprehensions as shown above to help, also other filters and lookups can be chained and leveraged to achieve more complex transformations.
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+Create dictionary from list
+---------------------------
+
+In most languages it is easy to create a dictionary (aka map/associatve array/hash/etc) from a list of pairs, in Ansible there are a couple of ways to do it and the best one for you might depend on the source of your data.
```
```suggestion
In most languages it is easy to create a dictionary (aka map/associative array/hash/etc) from a list of pairs, in Ansible there are multiple ways to do this and the best one for you might depend on the data source.
```
or
```suggestion
In most languages it is easy to create a dictionary (aka map/associative array/hash/etc) from a list of pairs, in Ansible there are a couple of ways to do it and the best one for you might depend on the source of your data.
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+      myvarnames: "{{q('varnames', '^my')}}"
+      mydict: "{{ dict(myvarnames | zip(q('vars', \*myvarnames))) }}"
+
+A quick explanation, since there is a lot to unpack from those 2 lines
```
```suggestion
A quick explanation, since there is a lot to unpack from these 2 lines
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+
+A quick explanation, since there is a lot to unpack from those 2 lines
+
+ - The 'varnames' lookup returns a list of variables that matches 'begin with "my"'
```
```suggestion
 - The 'varnames' lookup returns a list of variables that match 'begin with "my"'
```

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+A quick explanation, since there is a lot to unpack from those 2 lines
+
+ - The 'varnames' lookup returns a list of variables that matches 'begin with "my"'
+ - When feedng the list from the prevuois step into the 'vars' lookup to get the list of values.
```
```suggestion
 - When feeding the list from the previous step into the 'vars' lookup to get the list of values.
```
or should it be
```suggestion
 - Than the list from the previous step is feed into the 'vars' lookup to get the list of values.
```

felixfontein replied:
In the second suggestion, it should be `Then` and not `Than`.

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+
+ - The 'varnames' lookup returns a list of variables that matches 'begin with "my"'
+ - When feedng the list from the prevuois step into the 'vars' lookup to get the list of values.
+   The ``*`` is used to 'dereference the list' (a pythonism that works in Jinja), otherwise it would take the list as one argument.
```
```suggestion
   The ``*`` is used to 'dereference the list' (a pythonism that works in Jinja), otherwise it would take the list as a single argument.
```

agowa reviewed (changes requested):
Thanks for your contribution this already helped me with building better readable roles and playbooks ;-)

agowa commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,205 @@
+    - name: enable a list of windows features, by name
+      win_feature:
+        # here we filter out any 'omit' tokens we find in the supplied list
+        name: '{{ namestuff|reject('equal_to', omit)|list }}'
```
```suggestion
        name: '{{ namestuff|reject("equalto", omit)|list }}'
```
or
```suggestion
        name: "{{ namestuff|reject('equalto', omit)|list }}"
```

JonTheNiceGuy commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,214 @@
+  - tests: used to validate data, used with the ``is`` Jinja2 operator.
+
+.. _note:
+   * Some tests and filters are provided directly by Jinja2, so their aviailability depends on the Jinja2 version, not Ansible.
```
`aviailability` = `availability`

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,214 @@
+
+There are several ways to do it in Ansible, this is just one example:
+
+.. code-block:: YAML+Jinja2
```
```suggestion
.. code-block:: YAML+Jinja
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,214 @@
+In this case, we want to find the mount point for a given path across our machines, since we already collect mount facts, we can use the following:
+
+.. code-block::
+ :caption: Use selectattr to filter mounts into list i can then sort and select the last from
```
```suggestion
 :caption: Use selectattr to filter mounts into list I can then sort and select the last from
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,214 @@
+                      version: 1.1
+
+
+.. code-block::
```
I think you explicitly need to specify the language for the code-block if you have more parameters (like `:caption:` and `:emphasize-lines:`). I think that's causing (most of) the errors.

jctanner commented:
!waffling stale_ci

angystardust commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,214 @@
+This results in ``{"foo": "a", "var": "b", "bar": "c"}``.
+
+
+You can even combine these simple examples with other filters and looups to create a dictionary dynamically by matching patterns to variable names:
```
smal typo: looups -> lookups

angystardust commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,214 @@
+
+  vars:
+      single_list: [ 'a', 'b', 'c', 'd' ]
+      mydict: "{{ dict(singlelist) }}"
```
There's a missing underscore in the variable name. It should be:
`mydict: "{{ dict(single_list) }}"`

mgedmin replied:
You speak about a list of pairs, but that would be `[('a', 'b'), ('c', 'd')]`.

`dict(['a', 'b', 'c', 'd'])` is an error in Python:

> ValueError: dictionary update sequence element #0 has length 1; 2 is required

bcoca commented:
from @flowersong `host_most_recently_rebooted = ansible_play_hosts_all | map('extract', hostvars, 'ansible_uptime_seconds') | sort | first`

mkrizek commented:
!waffling ci_verified

bmillemathias commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+The Python equivalent code would be:
+
+.. code-block:: python
+
```
for the completeness, I'd add a line with what contains `chains` on all examples down there

bmillemathias commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+.. _complex_data_manipulation:
+
+Complex data manipulation
```
"Advanced" instead of "Complex" ? for me this is complex (because I'm not used to it) and advanced (because I never use that)

Akasurde replied:
I think data is complex and syntax is advance.

bmillemathias commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+Loops and list comprehensions
+=============================
+
+Most programmers are used to use loops (``for``, ``while``, etc.) and list comprehensions to do transformations on lists including lists of objects, Jinja2 has a few filters that provide this functionality: ``map``, ``select``, ``reject``, ``selectattr``, ``rejectattr``.
```
Are you sure the audience are only programmers ? I like to think the power of Ansible is to widen the audience than developers to sysadmin as well, I'm a sysadmin and while I can write bash or python script (I know list comprehension but not sure every people reading are aware of what it is) I would not call me programmer.

If this first part is to do an analogy with "standard" programming language, we could write that differently, like ``programming languages have loops .... lists of objects``

Akasurde replied:
Done.

bmillemathias commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+      vars:
+        namestuff:
+          - "{{ (fs_installed_smb_v1 | default(False)) | ternary(omit, 'FS-SMB1') }}"
+          - "foo"
```
That would help to put real features, so someone who want to test, can just make a copy-paste. (I don't know windows so can't help)

bmillemathias replied:
real features that does not break things by the way :)

bmillemathias reviewed (commented):
Overall a very good document, I see the real value of it and learn stuff.

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+.. _complex_data_manipulation:
+
+data manipulation
```
```suggestion
Data manipulation
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+Loops and list comprehensions
+=============================
+
+Most programming languages have loops (``for``, ``while``, etc.) and list comprehensions to do transformations on lists including lists of objects, Jinja2 has a few filters that provide this functionality: ``map``, ``select``, ``reject``, ``selectattr``, ``rejectattr``.
```
```suggestion
Most programming languages have loops (``for``, ``while``, etc.) and list comprehensions to do transformations on lists including lists of objects. Jinja2 has a few filters that provide this functionality: ``map``, ``select``, ``reject``, ``selectattr``, ``rejectattr``.
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+
+Most programming languages have loops (``for``, ``while``, etc.) and list comprehensions to do transformations on lists including lists of objects, Jinja2 has a few filters that provide this functionality: ``map``, ``select``, ``reject``, ``selectattr``, ``rejectattr``.
+
+  - map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on attributes of the list elements.
+  - select/reject: this is a for loop with a condition, that allows you to create a subset of a list that matches (or not) based on the result of the condition.
+  - selectattr/rejectattr: very similar to the above but it uses a specific attribute of the list elements for the conditional statement.
```
```suggestion
- map: this is a basic for loop that just allows you to change every item in a list, using the 'attribute' keyword you can do the transformation based on attributes of the list elements.
- select/reject: this is a for loop with a condition, that allows you to create a subset of a list that matches (or not) based on the result of the condition.
- selectattr/rejectattr: very similar to the above but it uses a specific attribute of the list elements for the conditional statement.
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+                      version: 1.1
+
+
+.. code-block:: none
```
```suggestion
.. code-block:: ansible-output
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+Another way is to avoid adding elements to the list in the first place, so you can just use it directly:
+
+.. code-block:: YAML+Jinja
+ :caption: using set_fact in a loop to increment a list conditionally
```
```suggestion
 :caption: Using set_fact in a loop to increment a list conditionally
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+Complex Type transformations
+=============================
+
+Jinja provides filters for simple data type transformations (int, bool, etc), but when you want to transform data structures things are not as easy.
```
```suggestion
Jinja provides filters for simple data type transformations (``int``, ``bool``, etc), but when you want to transform data structures things are not as easy.
```

felixfontein replied:
This one's missing.

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+
+
+.. code-block:: YAML+Jinja
+ :caption: its simpler when it is a list of pairs:
```
```suggestion
 :caption: It is simpler when we have a list of pairs:
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+        myvarnames: "{{ q('varnames', '^my') }}"
+        mydict: "{{ dict(myvarnames | zip(q('vars', *myvarnames))) }}"
+
+A quick explanation, since there is a lot to unpack from these 2 lines
```
```suggestion
A quick explanation, since there is a lot to unpack from these two lines:
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+
+A quick explanation, since there is a lot to unpack from these 2 lines
+
+ - The 'varnames' lookup returns a list of variables that match 'begin with "my"'
```
```suggestion
 - The ``varnames`` lookup returns a list of variables that match "begin with ``my``".
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+A quick explanation, since there is a lot to unpack from these 2 lines
+
+ - The 'varnames' lookup returns a list of variables that match 'begin with "my"'
+ - When feeding the list from the previous step into the 'vars' lookup to get the list of values.
```
```suggestion
 - Then feeding the list from the previous step into the ``vars`` lookup to get the list of values.
```

felixfontein commented on docs/docsite/rst/user_guide/complex_data_manipulation.rst:
```diff
@@ -0,0 +1,243 @@
+ - The 'varnames' lookup returns a list of variables that match 'begin with "my"'
+ - When feeding the list from the previous step into the 'vars' lookup to get the list of values.
+   The ``*`` is used to 'dereference the list' (a pythonism that works in Jinja), otherwise it would take the list as a single argument.
+ - Both lists get passed to the 'zip' filter to pair them off into a unified list (key, value, key2, value2, ...)
```
```suggestion
 - Both lists get passed to the ``zip`` filter to pair them off into a unified list (key, value, key2, value2, ...).
```

bcoca committed:
Update docs/docsite/rst/user_guide/complex_data_manipulation.rst

Co-authored-by: Felix Fontein <felix@fontein.de>

felixfontein reviewed (approved):
LGTM
